### DISCUSSION TITLE
Feature request : Carousel

### DISCUSSION THREAD

**Discussion by nisalk96** [mantinedev/mantine#313] [Feedback]
Mantine is beautiful 😍 modern UI library 
It would be better if it's have a Carousel/Slide show component (feature request)
with horizontal and vertical nav 😎
Eg:
1. https://ant.design/components/carousel/#header
2. https://getuikit.com/docs/slideshow

**Comment by rtivital** [comment]
Hi @Nzent, thanks for feedback, we will be adding carousel in one of 3.x releases. Can you also share features that are important for you in Carousel component?

**Reply by Shibly-Noman** [reply]
@rtivital 
Mantine is awesome. I just found out it recently one of fireship.io's video. 
One thing I just want to add with @Nzent that,
Using Owl or Slick carousel with react is real pain. Owl carousel still has dependency with jQuery and both are real hard to customize. 
One feature is really important to me (and I think most other front-end'ers) that **dynamically call API on next/previous navigation**.  To be more specific **Carousel with Dynamic Pagination**.

**Reply by ajnart** [reply]
> Hi @Nzent, thanks for feedback, we will be adding carousel in one of 3.x releases. Can you also share features that are important for you in Carousel component?

Any updates on ETA for a carousel component ?

**Reply by rtivital** [reply]
No, we've changed the plans, Carousel is not the priority.

**Comment by nisalk96** [comment]
@rtivital 
1. Switch slides with arrow navigation (Left, Right)
2. Slides direction (Verticle, Horizontal)
3. Image thumb nav like eBay items slider
4. Dots indicator
5. Slide transition animations (fade, slide)
6. Slide interval
7. Slides autoplay
8. Looping slides

**Comment by itsmichaeldiego** [comment]
How are we looking into the Carousel component? Would love to use mantine's!!

**Comment by jordanpinski** [comment]
+1 would love to use mantine's carousel. Hoping this can be added back into the roadmap.

**Comment by giffordgshane** [comment]
+1 would love to see this added to mantine's arsenal.

**Comment by alexandraj777** [comment]
+1 For a mantine carousel!

**Comment by nisalk96** [comment]
@rtivital thank you very much for adding this feature to the Mantine library💖 it's a very useful component 🔥😍😍

**Reply by ajnart** [reply]
Mark the discussion as solved :)

**Reply by webdiller** [reply]
How can I change current slide with API?

**Comment by sandromileno** [comment]
Is there a way to control the navigation via methods?
like this:
https://ant.design/components/carousel/#header

**Reply by rtivital** [reply]
https://www.embla-carousel.com/api/methods/

**Comment by MadaShindeInai** [comment]
How can i implement autoplay opportunity with Carousel?

**Reply by giffordgshane** [reply]
You can do this by using embla autoplay.

```
import { useRef } from 'react';
import Autoplay from 'embla-carousel-autoplay';
import { Carousel } from '@mantine/carousel';

function Demo() {
  const autoplay = useRef(Autoplay({ delay: 2000 }));
  return (
    <Carousel
      maw={320}
      mx="auto"
      withIndicators
      height={200}
      plugins={[autoplay.current]}
      onMouseEnter={autoplay.current.stop}
      onMouseLeave={autoplay.current.reset}
    >
      <Carousel.Slide>1</Carousel.Slide>
      <Carousel.Slide>2</Carousel.Slide>
      <Carousel.Slide>3</Carousel.Slide>
      {/* ...other slides */}
    </Carousel>
  );
}
```

**Reply by MadaShindeInai** [reply]
Thank you! Saw "plugins" prop before asking, but failed to add there Autoplay.

**Reply by PaulPintang** [reply]
> You can do this by using embla autoplay.
> 
> ```
> import { useRef } from 'react';
> import Autoplay from 'embla-carousel-autoplay';
> import { Carousel } from '@mantine/carousel';
> 
> function Demo() {
>   const autoplay = useRef(Autoplay({ delay: 2000 }));
>   return (
>     <Carousel
>       maw={320}
>       mx="auto"
>       withIndicators
>       height={200}
>       plugins={[autoplay.current]}
>       onMouseEnter={autoplay.current.stop}
>       onMouseLeave={autoplay.current.reset}
>     >
>       <Carousel.Slide>1</Carousel.Slide>
>       <Carousel.Slide>2</Carousel.Slide>
>       <Carousel.Slide>3</Carousel.Slide>
>       {/* ...other slides */}
>     </Carousel>
>   );
> }
> ```

this code from mantine docs is not working. don't know why.

**Reply by giffordgshane** [reply]
Possible reasons for the Carousel code not working:

- Missing dependencies: Make sure to install both the @mantine/carousel and embla-carousel-autoplay packages as project dependencies.
- Incorrect import: Verify that the Autoplay component is imported from 'embla-carousel-autoplay'.
- Compatibility issues: Ensure that the versions of @mantine/carousel and embla-carousel-autoplay packages are compatible with each other.
- Check for errors: Look for error messages in the browser console or terminal for clues about the issue.

Feel free to provide more specific details or error messages, and I'll be glad to help you further.

**Reply by PaulPintang** [reply]
thanks @giffordshane . fix my issue by downgrading the embla-autoplay plugin.

**Reply by aarav-classiebit** [reply]
Which version of embla-autoplay plugin are you using?

**Reply by SaiSugeethKamineni** [reply]
I had the same issue. Upon checking package.json, I discovered that the latest embla-carousel-autoplay 8.0.0-rc05 was installed instead of the stable 7.1.0 version. Downgrading it to the stable version fixed the issue.

**Reply by AshutoshDash1999** [reply]
This worked. Make sure to use `npm i` after changing the version and restart the server.

**Reply by AC-Agiir** [reply]
Hello, I'm working on a React-Typescript project, and I have a type error when I'm passing `[autoplay.current]` in props `plugins` of Carousel, just like the documentation :

`Type 'AutoplayType' is not assignable to type 'CreatePluginType<LoosePluginType, {}>'.`

Here is my component `MyCarousel.tsx` :

```
import React, {useRef} from "react";
import Autoplay from 'embla-carousel-autoplay';
import { Carousel } from '@mantine/carousel';

type Props = {}

export const MyCarousel: React.FC<Props> = () => {
    const autoplay = useRef(Autoplay({ delay: 2000 }));

    return (
            <Carousel
                mx="auto"
                my="md"
                height={300}
                withIndicators
                plugins={[autoplay.current]}
                onMouseEnter={autoplay.current.stop}
                onMouseLeave={autoplay.current.reset}
            >
                {/* ...slides */}
            </Carousel>
    )
}
```

My dependencies in package.json :

```
"dependencies": {
	"@mantine/carousel": "^6.0.16",
	"@mantine/core": "^6.0.16",
	"@mantine/form": "^6.0.16",
	"@mantine/hooks": "^6.0.16",
	"@mantine/styles": "^6.0.16",
	"@tabler/icons-react": "^2.24.0",
	"clay": "^2.24.1",
	"embla-carousel-autoplay": "^7.1.0",
	"liferay-module-config-generator": "^1.3.3",
	"react": "^18.2.0",
	"react-dom": "^18.2.0",
	"react-icons-kit": "^2.0.0",
	"react-redux": "^8.1.1",
	"react-router-dom": "^6.14.1",
	"react-transition-group": "^4.4.5",
	"redux-thunk": "^2.4.2"
},
```

I tried downgrading `embla-carousel-autoplay` to stable version 6, but it doesn't solve my problem, any ideas?

**Reply by Akbaroke** [reply]
I succeeded after using the code and packege `"embla-carousel-autoplay": "^7.1.0",`

**Comment by egl74** [comment]
Is there a way to horizontally scroll the Carousel with Magic Mouse/Trackpad?

**Reply by giffordgshane** [reply]
```
import { Carousel } from '@mantine/carousel';

function Demo() {
  return (
    <Carousel dragFree>
      {/* ...slides */}
    </Carousel>
  );
}
```

With this configuration, you should be able to scroll horizontally through the slides by swiping left or right on the Magic Mouse or Trackpad. The slides will smoothly follow the swipe gesture without snapping to specific positions.

**Comment by Asoft123** [comment]
Autoplay with Mantine using  embla-carousel-autoplay does not work for me

**Reply by Asoft123** [reply]
How it was used

import { Carousel } from "@mantine/carousel"
	
const autoplay = useRef(Autoplay({delay:200}));

plugins={[autoplay.current ]}

**Reply by Bidhan-Bajracharya** [reply]
Assuming you were using TS, I was getting an error as well. The error was as mentioned in above discussion: 
`Type 'AutoplayType' is not assignable to type 'CreatePluginType<LoosePluginType, {}>'`

I solved it by downgrading the version of `embla-carousel-autoplay` to `"^7.1.0"`. After installing this version and restarting the server, the error was gone and everything worked as expected. Here is the code snippet:

```
import { Carousel } from '@mantine/carousel';
import Autoplay from 'embla-carousel-autoplay';
import { useRef } from 'react';

const TrustedBy = () => {
  const autoplay = useRef(Autoplay({ delay: 2000 }));
  return (
    <>
      <section className="flex min-h-[50vh] min-w-full flex-col items-center overflow-x-clip  bg-[#F6F6F6] pt-5 dark:bg-[#253e49]">
        <div className="w-full overflow-hidden">
          <Carousel
            height={310}
            slideSize={{ base: '100%', sm: '50%', md: '25.333333%' }}
            slideGap={{ base: 0, sm: 'md' }}
            align="start"
            loop
            dragFree
            withControls={false}
            plugins={[autoplay.current]}
            onMouseEnter={autoplay.current.stop}
            onMouseLeave={autoplay.current.reset}
          >
            {...slides}
          </Carousel>
        </div>
      </section>
    </>
  );
};
```

**Comment by due23** [comment]
Is there a way to change the value of the withControls and draggable options for different breakpoints like you can do with slideSize. I want to be able to disable/hide the controls when there are not enough slides to fill the number of slides the carousel allows at each breakpoint.

Ideally I would like to do something like the following but it doesn't seem possible given withControls and draggable only take one boolean value....

```
function Demo() {
  const slidesList = [1, 2, 3, 4];
  return (
    <Carousel
    slideSize={{
      base: "33.333333%", // show 3 slides at a time
      xs: "25%", // show 4 slides at a time
      sm: "20%", // show 5 slides at a time
      md: "16.66666666%", // show 6 slides at a time
    }}
    maw={{
      base: "306px",
      xs: "376px",
      sm: "446px",
      md: "516px",
    }}
    withControls={{
      base: slidesList.length <= 3 ? false : true,
      xs: slidesList.length <= 4 ? false : true,
      sm: slidesList.length <= 5 ? false : true,
      md: slidesList.length <= 6 ? false : true,
    }}
    align="start"
    loop
    pr="3em"
    pl="3em"
  >
    {slidesList.map((slide) => (
      <Carousel.Slide>{slide}</Carousel.Slide>
    ))}
  </Carousel>
  
  );
}
```
